README.md
# docker-nginx

[![Build Status](https://travis-ci.com/madharjan/docker-nginx.svg?branch=master)](https://travis-ci.com/madharjan/docker-nginx)

Docker container for Nginx based on [madharjan/docker-base](https://github.com/madharjan/docker-base/)

## Features

* Nginx error log forwarded to Docker logs
* Bats [bats-core/bats-core](https://github.com/bats-core/bats-core) based test cases
* Deploy/update web projects from git
* Setup reverse proxy

## Nginx 1.18.0 (docker-nginx)

### Environment

| Variable         | Default | Example                                                           |
| ---------------- | ------- | ----------------------------------------------------------------- |
| DISABLE_NGINX    | 0       | 1 (to disable)                                                    |
|                  |         |                                                                   |
| INSTALL_PROJECT  | 0       | 1 (to enable)                                                     |
| PROJECT_GIT_REPO |         | <https://github.com/BlackrockDigital/startbootstrap-creative.git> |
| PROJECT_GIT_TAG  | HEAD    | v5.1.4                                                            |
|                  |         |                                                                   |
| DEFAULT_PROXY    | 0       | 1 (to enable)                                                     |
| PROXY_SCHEME     | http    | https                                                             |
| PROXY_HOST       |         | 127.0.0.1                                                         |
| PROXY_PORT       | 8080    | 8000                                                              |

## Build

```bash
# clone project
git clone https://github.com/madharjan/docker-nginx
cd docker-nginx

# build
make

# tests
make run
make test

# clean
make clean
```

## Run

**Note**: update environment variables below as necessary

```bash
# prepare foldor on host for container volumes
sudo mkdir -p /opt/docker/nginx/etc/conf.d
sudo mkdir -p /opt/docker/nginx/html/
sudo mkdir -p /opt/docker/nginx/log/

# stop & remove previous instances
docker stop nginx
docker rm nginx

# run container
docker run -d \
  -p 80:80 \
  -v /opt/docker/nginx/etc/conf.d:/etc/nginx/conf.d \
  -v /opt/docker/nginx/html:/var/www/html \
  -v /opt/docker/nginx/log:/var/log/nginx \
  --name nginx \
  madharjan/docker-nginx:1.18.0
```

## Systemd Unit File

**Note**: update environment variables below as necessary

```txt
[Unit]
Description=Nginx

After=docker.service

[Service]
TimeoutStartSec=0

ExecStartPre=-/bin/mkdir -p /opt/docker/nginx/etc/conf.d
ExecStartPre=-/bin/mkdir -p /opt/docker/nginx/html
ExecStartPre=-/bin/mkdir -p /opt/docker/nginx/log
ExecStartPre=-/usr/bin/docker stop nginx
ExecStartPre=-/usr/bin/docker rm nginx
ExecStartPre=-/usr/bin/docker pull madharjan/docker-nginx:1.18.0

ExecStart=/usr/bin/docker run \
  -p 80:80 \
  -v /opt/docker/nginx/etc/conf.d:/etc/nginx/conf.d \
  -v /opt/docker/nginx/html:/var/www/html \
  -v /opt/docker/nginx/log:/var/log/nginx \
  --name nginx \
  madharjan/docker-nginx:1.18.0

ExecStop=/usr/bin/docker stop -t 2 nginx

[Install]
WantedBy=multi-user.target
```

## Generate Systemd Unit File

| Variable         | Default     | Example                                                           |
| ---------------- | ----------- | ----------------------------------------------------------------- |
| PORT             |             | 8080                                                              |
| VOLUME_HOME      | /opt/docker | /opt/data                                                         |
| NAME             | ngnix       |                                                                   |
|                  |             |                                                                   |
| INSTALL_PROJECT  | 0           | 1 (to enable)                                                     |
| PROJECT_GIT_REPO |             | <https://github.com/BlackrockDigital/startbootstrap-creative.git> |
| PROJECT_GIT_TAG  | HEAD        | v5.1.4                                                            |
|                  |             |                                                                   |
| DEFAULT_PROXY    | 0           | 1 (to enable)                                                     |
| PROXY_SCHEME     | http        | https                                                             |
| PROXY_HOST       |             | 127.0.0.1                                                         |
| PROXY_PORT       | 8080        | 8000                                                              |
|                  |             |                                                                   |
| LINK_CONTAINERS  |             | nginx-web2py:app,nginx:website                                    |

### With deploy web projects

```bash
# generate nginx.service
docker run --rm \
  -e PORT=80 \
  -e INSTALL_PROJECT=1 \
  -e PROJECT_GIT_REPO=https://github.com/BlackrockDigital/startbootstrap-creative.git \
  -e PROJECT_GIT_TAG=v5.1.4 \
  madharjan/docker-nginx:1.18.0 \
  nginx-systemd-unit | \
  sudo tee /etc/systemd/system/nginx.service

sudo systemctl enable nginx
sudo systemctl start nginx
```

### With reverse proxy

```bash
# generate nginx.service
docker run --rm \
  -e PORT=80 \
  -e DEFAULT_PROXY=1 \
  -e PROXY_HOST=odoo \
  -e PROXY_PORT=8080 \
  -e LINK_CONTAINERS=odoo:odoo,nginx:website \
  madharjan/docker-nginx:1.18.0 \
  nginx-systemd-unit | \
  sudo tee /etc/systemd/system/nginx.service

sudo systemctl enable nginx
sudo systemctl start nginx
```

## Add virtualhost reverse proxy config

| Variable         | Default | Example     |
| ---------------- | ------- | ----------- |
| PROXY_VHOST_NAME |         | myapp.local |
| PROXY_SCHEME     | http    | https       |
| PROXY_HOST       |         | 127.0.0.1   |
| PROXY_PORT       | 8080    | 8000        |

```bash
# add proxy.conf
docker exec -it \
  -e PROXY_VHOST_NAME=myapp \
  -e PROXY_HOST=172.18.0.5 \
  -e PROXY_PORT=8080 \
  nginx \
  nginx-vhost-proxy-conf
```


Changelog.md

# Changelog

## 1.18.0

* Updated to be based on madharjan/docker-base:20.04
* Using Nginx 1.18.0

## 1.10.3

* Updated to be based on madharjan/docker-base:16.04
* Added error_page config 50x,403,404,405 and test cases
* Using Nginx 1.10.3
* Added web project deploy
* Added reverse proxy setup

## 1.4.6

* Based on madharjan/docker-base:14.04
* Using Nginx 1.4.6


scripts/install.sh
#!/bin/bash
set -e
export LC_ALL=C
export DEBIAN_FRONTEND=noninteractive

if [ "${DEBUG}" = true ]; then
  set -x
fi

NGINX_CONFIG_PATH=/build/config/nginx

apt-get update

## Install Project
apt-get install -y --no-install-recommends git-core

cp /build/services/15-project.sh /etc/my_init.d
chmod 750 /etc/my_init.d/15-project.sh

## Install Nginx and runit service
/build/services/nginx/nginx.sh

mkdir -p /config/etc/nginx/conf.d
mkdir -p /config/var/www/html

cp ${NGINX_CONFIG_PATH}/default.conf /config/etc/nginx/conf.d/default.conf
cp ${NGINX_CONFIG_PATH}/default-proxy.conf /config/etc/nginx/conf.d/default-proxy.conf
cp ${NGINX_CONFIG_PATH}/vhost-proxy.conf /config/etc/nginx/conf.d/vhost-proxy.conf
cp ${NGINX_CONFIG_PATH}/index.html /config/var/www/html/index.html
cp ${NGINX_CONFIG_PATH}/403.html /config/var/www/html/403.html
cp ${NGINX_CONFIG_PATH}/404.html /config/var/www/html/404.html
cp ${NGINX_CONFIG_PATH}/405.html /config/var/www/html/405.html
cp ${NGINX_CONFIG_PATH}/50x.html /config/var/www/html/50x.html
cp ${NGINX_CONFIG_PATH}/favicon.ico /config/var/www/html/favicon.ico

mkdir -p /etc/my_init.d
cp /build/services/20-nginx.sh /etc/my_init.d
chmod 750 /etc/my_init.d/20-nginx.sh

cp /build/bin/nginx-systemd-unit /usr/local/bin
chmod 750 /usr/local/bin/nginx-systemd-unit
cp /build/bin/nginx-vhost-proxy-conf /usr/local/bin
chmod 750 /usr/local/bin/nginx-vhost-proxy-conf



services/20-nginx.sh
#!/bin/bash

set -e

if [ "${DEBUG}" = true ]; then
  set -x
fi

DISABLE_NGINX=${DISABLE_NGINX:-0}
DEFAULT_PROXY=${DEFAULT_PROXY:-0}
PROXY_HOST=${PROXY_HOST:-}
PROXY_PORT=${PROXY_PORT:-8080}
PROXY_SCHEME=${PROXY_SCHEME:-http}

if [ ! "${DISABLE_NGINX}" -eq 0 ]; then
  touch /etc/service/nginx/down
  touch /etc/service/nginx-log-forwarder/down
else
  rm -f /etc/service/nginx/down
  rm -f /etc/service/nginx-log-forwarder/down
fi

if [ -f /etc/nginx/conf.d/default.conf ]; then
   echo "Nginx config 'default.conf' already exists"
else
  if [ ! "${DEFAULT_PROXY}" -eq 0 ]; then
    if [ -n "${PROXY_HOST}" ]; then
      cp /config/etc/nginx/conf.d/default-proxy.conf /etc/nginx/conf.d/default.conf

      export PROXY_HOST PROXY_PORT PROXY_SCHEME
      perl -p -i -e 's/\$\{([^}]+)\}/defined $ENV{$1} ? $ENV{$1} : $&/eg' /etc/nginx/conf.d/default.conf

    fi
  else 
    cp /config/etc/nginx/conf.d/default.conf /etc/nginx/conf.d/default.conf
  fi
fi

if [ -f /var/www/html/index.html ]; then
   echo "Nginx content already exists"
else
  mkdir -p /var/www/html
  cp /config/var/www/html/index.html /var/www/html/index.html
  cp /config/var/www/html/403.html /var/www/html/403.html
  cp /config/var/www/html/404.html /var/www/html/404.html
  cp /config/var/www/html/405.html /var/www/html/405.html
  cp /config/var/www/html/50x.html /var/www/html/50x.html
  cp /config/var/www/html/favicon.ico /var/www/html/favicon.ico

  chown -R www-data:www-data /var/www/html/*
  touch /var/www/html/.403test
  chmod 640 /var/www/html/.403test
fi
